lib/jxl/encode_test.cc

```diff
@@ -1022,6 +1023,12 @@ TEST(EncodeTest, AnimationHeaderTest) {
   // Decode to verify the boxes, we don't decode to pixels, only the boxes.
   JxlDecoderPtr dec = JxlDecoderMake(nullptr);
   EXPECT_NE(nullptr, dec.get());
+
+  // To test the blend_info fields, coalescing must be set to false in the
+  // decoder. This also assumes the encoder keeps them as-is, in theory it's
+  // permitted for the encoder to alter the blend mode as long as it gives
+  // the same result when coalescing.
```

Comment:
This is debatable imo: do we expect an encoder to (visually) preserve the coalesced frames or the non-coalesced frames?

I think generally for animations you mostly care only about the coalesced frames and it's fine if the encoder chooses its own blend modes and crops, while for layered still images I think the blend mode and crop is part of the image metadata that you want to preserve. In particular, it can be useful to have layers that are larger than the canvas and I would consider it wrong for an encoder to silently trim it to canvas size (which would of course still give the same result when coalescing).

I think in the end we might need an encode option to select whether the encoder is allowed to alter crops and/or blend modes. In general there might be a few of these things where we may want the encoder to take more or less liberty in changing things, like e.g. dropping a trivial (all-opaque) alpha channel, silently convert RGB to Grayscale if R=G=B, auto-adjust the bit depth (e.g. if all pixels are 0 or 255), remove duplicate frames and add up their durations, preserve invisible pixels or not, etc etc. It is use-case dependent where the boundary lies between what an encoder is allowed to do and what it isn't allowed to do.

Reply:
Makes sense, I've removed this comment for now.

For next steps, if you don't explicitely call JxlEncoderFrameSettingsSetInfo I'd say the encoder should be able to choose (to e.g. split one frame into multiple internal frames if some specific settings are enabled), and when explicitely calling it, indeed options to enable or disable encoding techniques that do this can be added

Reply:
Splitting a frame into multiple internal frames that are not kRegularFrame is always fine imo - but changing what you get when doing a non-coalesced decode is something else.